elk/utils/typing.py

```diff
@@ -10,3 +12,23 @@ def assert_type(typ: Type[T], obj: Any) -> T:
         raise TypeError(f"Expected {typ.__name__}, got {type(obj).__name__}")
 
     return cast(typ, obj)
+
+
+def float32_to_int16(x: torch.Tensor) -> torch.Tensor:
+    """Converts float32 to float16, then reinterprets as int16."""
+    return x.type(torch.float16).view(torch.int16)
+
+
+def int16_to_float32(x: torch.Tensor) -> torch.Tensor:
+    """Converts int16 to float16, then reinterprets as float32."""
+    return x.view(torch.float16).type(torch.float32)
+
+
+def upcast_hiddens(hiddens: torch.Tensor) -> torch.Tensor:
```

Comment:
This function definitely isn't needed, right? You're just calling `int16_to_float32` and _redundantly_ asserting the type of the hiddens (which already have to be a `torch.Tensor` since you specified that in the type signature of the function)